core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/RetryableStage2.java

```diff
@@ -83,8 +94,52 @@ private Response<OutputT> executeRequest(RetryableStageHelper2 retryableStageHel
         retryableStageHelper.setLastResponse(response.httpResponse());
         if (!response.isSuccess()) {
             retryableStageHelper.adjustClockIfClockSkew(response);
-            throw response.exception();
+            throw responseException(response);
         }
         return response;
     }
+
+    private RuntimeException responseException(Response<OutputT> response) {
+        Optional<Integer> optionalRetryAfter = retryAfter(response.httpResponse());
+        if (optionalRetryAfter.isPresent()) {
+            return new SdkExceptionWithRetryAfterHint(optionalRetryAfter.get(), response.exception());
+        }
+        return response.exception();
+    }
+
+    private Optional<Integer> retryAfter(SdkHttpFullResponse response) {
+        Optional<String> optionalRetryAfterHeader = response.firstMatchingHeader(RETRY_AFTER_HEADER);
+        if (optionalRetryAfterHeader.isPresent()) {
+            String retryAfterHeader = optionalRetryAfterHeader.get();
+            try {
+                return Optional.of(Integer.parseInt(retryAfterHeader));
+            } catch (NumberFormatException e) {
+                // Ignore and fallback to returning empty.
+            }
+        }
+        return Optional.empty();
+    }
+
+    // This probably should go directly into SdkException
```

Comment:
How do you propose to refactor?

Reply:
I'd think this can be part of  `SdkServiceException` with a new method that returns `Optional<Instant> retryAfterHint()` which defaults to `Optional.empty()`, and which can be added to the methods that we already use for retrying, e.g.,

```
    public boolean isClockSkewException() { … }
    public boolean isThrottlingException() { … }
    public Optional<Instant> retryAfterHint() { … }
```

Error handlers per protocol can take care of filling this up.

Reply:
I see. Agreed to handle this in offline design forum.